File: Network Delay Time - Leetcode 743/Network Delay Time - Leetcode 743.js
function networkDelayTime(times, n, k) {
    const graph = new Map();
    for (const [u, v, time] of times) {
        if (!graph.has(u)) graph.set(u, []);
        graph.get(u).push([v, time]);
    }

    const minHeap = new MinPriorityQueue();
    minHeap.enqueue([k, 0], 0);
    const minTimes = new Map();

    while (!minHeap.isEmpty()) {
        const [node, currentTime] = minHeap.dequeue().element;

        if (minTimes.has(node)) continue;
        minTimes.set(node, currentTime);

        if (graph.has(node)) {
            for (const [neighbor, time] of graph.get(node)) {
                if (!minTimes.has(neighbor)) {
                    minHeap.enqueue([neighbor, currentTime + time], currentTime + time);
                }
            }
        }
    }

    return minTimes.size === n ? Math.max(...minTimes.values()) : -1;
}

File: Network Delay Time - Leetcode 743/Network Delay Time - Leetcode 743.java
import java.util.*;

public class Solution {
    public int networkDelayTime(int[][] times, int n, int k) {
        Map<Integer, List<int[]>> graph = new HashMap<>();
        for (int[] time : times) {
            int u = time[0], v = time[1], t = time[2];
            graph.computeIfAbsent(u, x -> new ArrayList<>()).add(new int[]{v, t});
        }

        PriorityQueue<int[]> minHeap = new PriorityQueue<>(Comparator.comparingInt(a -> a[1]));
        minHeap.add(new int[]{k, 0});
        Map<Integer, Integer> minTimes = new HashMap<>();

        while (!minHeap.isEmpty()) {
            int[] current = minHeap.poll();
            int node = current[0];
            int time = current[1];

            if (minTimes.containsKey(node)) continue;

            minTimes.put(node, time);
            if (graph.containsKey(node)) {
                for (int[] neighbor : graph.get(node)) {
                    int nei = neighbor[0];
                    int neiTime = neighbor[1];
                    if (!minTimes.containsKey(nei)) {
                        minHeap.add(new int[]{nei, time + neiTime});
                    }
                }
            }
        }

        if (minTimes.size() == n) {
            return Collections.max(minTimes.values());
        } else {
            return -1;
        }
    }
}
